pkg/ccl/streamingccl/streamproducer/replication_stream_test.go

```diff
@@ -694,30 +726,31 @@ USE d;
 	// Range is t1e - t2sn, emitting t2s - t2sn.
 	require.NoError(t, h.SysServer.DB().DelRangeUsingTombstone(ctx, t1Span.EndKey, t2Span.Key.Next()))
 
-	// Expected DelRange spans after sorting.
-	expectedDelRangeSpan1 := roachpb.Span{Key: t1Span.Key, EndKey: t1Span.EndKey}
-	expectedDelRangeSpan2 := roachpb.Span{Key: t2Span.Key, EndKey: t2Span.EndKey}
-	expectedDelRangeSpan3 := roachpb.Span{Key: t2Span.Key, EndKey: t2Span.Key.Next()}
+	// Expected DelRange events. We store these and the received
+	// del ranges in maps to account for possible duplicate
+	// delivery.
+	expectedDelRanges := make(map[string]struct{})
+	expectedDelRanges[roachpb.Span{Key: t1Span.Key, EndKey: t1Span.EndKey}.String()] = struct{}{}
+	expectedDelRanges[roachpb.Span{Key: t2Span.Key, EndKey: t2Span.EndKey}.String()] = struct{}{}
+	expectedDelRanges[roachpb.Span{Key: t2Span.Key, EndKey: t2Span.Key.Next()}.String()] = struct{}{}
 
-	codec := source.mu.codec.(*partitionStreamDecoder)
-	receivedDelRanges := make([]kvpb.RangeFeedDeleteRange, 0, 3)
+	receivedDelRanges := make(map[string]struct{})
 	for {
 		source.mu.Lock()
 		require.True(t, source.mu.rows.Next())
 		source.mu.codec.decode()
 		if codec.e.Batch != nil {
-			receivedDelRanges = append(receivedDelRanges, codec.e.Batch.DelRanges...)
+			for _, dr := range codec.e.Batch.DelRanges {
+				receivedDelRanges[dr.Span.String()] = struct{}{}
+			}
 		}
 		source.mu.Unlock()
-		if len(receivedDelRanges) == 3 {
+		if len(receivedDelRanges) >= 3 {
```

Comment:
if we get multiple duplicates of a key, is it possible for this test to flake? i.e. we'll break even though we haven't received one of the keys?

Would it better to break once we have received 3 unique delranges?

Reply:
I think that is what this does because receivedDelRanges is a map where the key is the same of the del range.

Reply:
oh lol, I should have looked more closely. Yes, that's exactly what you're doing.